packages/zql-integration-tests/src/helpers/runner.ts

```diff
@@ -304,49 +317,188 @@ export async function createVitests<TSchema extends Schema>(
     }));
 }
 
-export async function runHydrationBenchmarks<TSchema extends Schema>(
-  {suiteName, zqlSchema, pgContent, only}: BenchOptions<TSchema>,
+type PushGenerator = (
+  iteration: number,
+) => [source: string, change: SourceChange][];
+
+export async function runBenchmarks<TSchema extends Schema>(
+  {suiteName, type, zqlSchema, pgContent, only}: HydrationOptions<TSchema>,
   ...benchSpecs: (readonly {
     name: string;
     createQuery: (q: Queries<TSchema>) => Query<TSchema, string>;
   }[])[]
-) {
+): Promise<void>;
+export async function runBenchmarks<TSchema extends Schema>(
+  {suiteName, type, zqlSchema, pgContent, only}: PushOptions<TSchema>,
+  ...benchSpecs: (readonly {
+    name: string;
+    createQuery: (q: Queries<TSchema>) => Query<TSchema, string>;
+    generatePush: PushGenerator;
+  }[])[]
+): Promise<void>;
+export async function runBenchmarks<TSchema extends Schema>(
+  {suiteName, type, zqlSchema, pgContent, only}: BenchOptions<TSchema>,
+  ...benchSpecs: (readonly {
+    name: string;
+    createQuery: (q: Queries<TSchema>) => Query<TSchema, string>;
+    generatePush?: PushGenerator;
+  }[])[]
+): Promise<void> {
   const dbs = await makeDatabases(suiteName, zqlSchema, pgContent);
   const delegates = makeDelegates(dbs, zqlSchema);
   const queries = makeQueries(zqlSchema, delegates);
 
-  return benchSpecs
+  benchSpecs
     .flat()
     .filter(t => (only ? only.includes(t.name) : true))
-    .map(({name, createQuery}) =>
+    .map(({name, createQuery, generatePush}) =>
       describe(name, () => {
-        makeHydrationBenchmark({
+        makeBenchmark({
+          zqlSchema,
+          delegates,
+          dbs,
           createQuery,
+          type,
           queries,
+          generatePush,
         });
       }),
     );
 }
 
-function makeHydrationBenchmark<TSchema extends Schema>({
+function makeBenchmark<TSchema extends Schema>({
+  zqlSchema,
   createQuery,
+  type,
   queries,
+  delegates,
+  dbs,
+  generatePush,
 }: {
+  zqlSchema: TSchema;
   createQuery: (q: Queries<TSchema>) => Query<TSchema, string>;
+  type: 'hydration' | 'push';
   queries: QueriesBySource<TSchema>;
+  delegates: Delegates;
+  dbs: DBs<TSchema>;
+  generatePush: PushGenerator | undefined;
 }) {
-  const zql = createQuery(queries.memory);
-  const zqlite = createQuery(queries.sqlite);
-  const pg = createQuery(queries.pg);
-  return [doBench('zql', zql), doBench('zqlite', zqlite), doBench('zpg', pg)];
+  switch (type) {
+    case 'push': {
```

Comment:
maybe makePushBenchmark and makeHydrationBenchmark since they do not share any code.